tokio/src/time/driver/entry.rs

```diff
@@ -1,361 +1,721 @@
-            waker: AtomicWaker::new(),
-            state: AtomicU64::new(state),
-            queued: AtomicBool::new(false),
-            error: AtomicU8::new(0),
-            next_atomic: UnsafeCell::new(ptr::null_mut()),
-            when: UnsafeCell::new(None),
-            next_stack: UnsafeCell::new(None),
-            prev_stack: UnsafeCell::new(ptr::null_mut()),
+    /// Transitions to the FiringInProgress state
+    pub(super) unsafe fn set_pending(&self) {
+        unsafe {
+            assert_eq!(
+                self.inner.as_ref().state.get(Ordering::Relaxed),
+                EntryState::Registered
+            );
+            self.inner
+                .as_ref()
+                .state
+                .set(EntryState::FiringInProgress, Ordering::Relaxed);
         }
     }
 
-    fn upgrade_inner(&self) -> Option<Arc<Inner>> {
-        self.inner.upgrade()
+    /// Attempts to transition to a terminal state. If the state is already a
+    /// terminal state, does nothing.
+    ///
+    /// Because the entry might be dropped immediately after the state is moved
+    /// to a terminal state, this function consumes the handle to ensure we
+    /// don't access the entry afterwards.
+    ///
+    /// Returns the last-registered waker, if any.
+    ///
+    /// SAFETY: The driver lock must be held while invoking this function.
+    pub(super) unsafe fn fire(self, completed_state: EntryState) -> Option<Waker> {
+        debug_assert!(completed_state.is_elapsed());
+
+        // Get a lifetime-erased pointer to the state. We'll use this to set the
+        // elapsed state below.
+        let pstate = unsafe { self.inner.as_ref() }.state.erase();
+
+        // We know this is currently registered... so its state should reflect
+        // this.
+        //
+        // Note that the state was set by another thread, but the mutex's
+        // implicit fences will synchronize this)
+        debug_assert!(matches!(
+            pstate.get(Ordering::Relaxed),
+            EntryState::Registered | EntryState::FiringInProgress
+        ));
+
+        pstate.set(EntryState::FiringInProgress, Ordering::Relaxed);
+
+        // Acq/rel barrier.
+        //
+        // We rely on the waker update/take acq/rel fences to ensure that this
+        // state write is not visible until after we have read any possible
+        // waker change (ie, any thread that might have changed the waker in
+        // between the state change and our taking the waker will definitely see
+        // the state change _after_ updating the waker.).
+        let waker: Option<Waker> = unsafe { self.inner.as_ref().waker.take_waker() };
+
+        // SAFETY: After this point we MUST NOT touch the referent of
+        // `self.inner`.
+        //
+        // Ordering is not critical for this. When we invoke the waker we will
+        // trigger an implicit fence to ensure that our state change is visible.
+        pstate.set(completed_state, Ordering::Relaxed);
```

Comment:
This was discussed offline, but for the record, this ordering is not sound as it can be re-ordered higher.

Reply:
Fixed in the commits I just pushed.